File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        # store a memo dictionary of the maximum points you can earn at each index
        memo = {}
        
        def maxPoint(index : int) -> int:
            if index in memo:
                return memo[index]
            
            # baseCase
            if index >= len(questions):
                return 0
            
            # we have 2 choices, solve or skip
            currPoint, currBrainPower = questions[index]

            solvePoint = currPoint + maxPoint(index + currBrainPower + 1)
            skipPoint = maxPoint(index+1)

            memo[index] = max(solvePoint, skipPoint)

            return memo[index]

        return maxPoint(0)            
        

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    // a hash map to store previosly calculated most points for indices
    Map<Integer, Long> memo = new HashMap();
    int[][] question;

    public long mostPointsForIndex(Integer index){

        // already precalculated value
        if (memo.containsKey(index)) {
            return memo.get(index);
        }
        
        // the base case
        if (index >= question.length){
            return 0;
        }

        long mostPointIfSolved = question[index][0] + mostPointsForIndex(index + question[index][1] + 1);
        long mostPointIfSkipped = mostPointsForIndex(index + 1);

        memo.put(index, Math.max(mostPointIfSolved, mostPointIfSkipped));

        return memo.get(index);
    }
    
    public long mostPoints(int[][] questions) {
        question = questions;
        return mostPointsForIndex(0);
    }
}